### ISSUE TITLE
Update traits.yml from APD

### ISSUE THREAD

**Issue by dfalster** [traitecoevo/austraits.build#684]
Hi @ehwenk 

This code should enable you to update the file `traits.yml` from the APD.csv

```{r}
library(tidyverse)
library(austraits.build)
```

First read and write the traits file to reformat it. Might want to add as a commit
```{r}
traits <- get_schema("config/traits.yml", I("traits"))
yaml::write_yaml(list(traits = traits), "config/traits.yml")
```

Then import data from APD
```{r}
APD <- read_csv("APD_traits.csv", show_col_types = FALSE) %>%
  filter(type_x == "numeric") %>%
  # make some updates 
  mutate(
    allowed_values_min=min, 
    allowed_values_max=max,
    description = paste0(description_encoded, ";", description),
    type= type_x,
    units = ifelse(str_starts(units, fixed("{")), sprintf("'%s'", units),units)
    )

traits <- get_schema("config/traits.yml", I("traits"))

for(i in nrow(APD)){
  trait_name <- APD$trait[i]
  if(!is.null(traits$elements[[trait_name]]))
    {
    for(v in c("units", "description", "label", "allowed_values_min", "allowed_values_max", "type", "comments")) {
      if(!is.na(APD[[v]][i]))
        traits$elements[[trait_name]][[v]] <- APD[[v]][i]
    }
  }
}

yaml::write_yaml(list(traits=traits), "config/traits.yml")
```

Using the above on the branch `trait-definitions-edits`, I only found a single trait where the defitiinon wasn't in sync:

![image](https://user-images.githubusercontent.com/3589590/235037615-b9873cf6-6f54-48cb-b3fe-e170ca1d7b46.png)

**Comment by ehwenk** [comment]
@dfalster 
I've expanded upon this to rebuild from scratch, but I don't know how to turn traits$elements$trait[i]$allowed_values_levels into a list for categorical traits, so the line `traits$elements$trait[i]$allowed_values_levels[j] <- util_df_to_list(categorical_tmp[j,])` fails

Otherwise it works well.

The file traits2.yml is simply
 `traits:
    description: Names and details of plant traits included in the AusTraits compilation
    type: list
    elements:  
    `
Other files are coming from APD

`library(tidyverse)
library(austraits.build)

APD <- read_csv("export/data/APD_traits.csv", show_col_types = FALSE) %>%
  mutate(
    description = ifelse(is.na(description), description_encoded, paste0(description_encoded, ";", description)),
    units = ifelse(str_starts(units, fixed("{")), sprintf("'%s'", units),units),
    Entity_URI = paste0("https://w3id.org/APD/traits/", identifier),
    allowed_values_levels = ifelse(type_x == "categorical", "add_in", NA),
  ) %>%
  select(trait,
         label,
         description,
         comments,
         type = type_x,
         units,
         allowed_values_min = min,
         allowed_values_max = max,
         allowed_values_levels,
         Entity_URI
         )

value_levels <- read_csv("export/data/APD_categorical_values.csv", show_col_types = FALSE) %>%
  select(trait_cat = trait_name,
         label,
         description
  )

traits <- get_schema("export/traits2.yml", I("traits"))

for(i in seq_along(1:nrow(APD))){
  trait_name <- APD$trait[i]
  {
    for(v in c("label", "description", "comments", "type", "units", "allowed_values_min", "allowed_values_max", "allowed_values_levels", "Entity_URI")) {
      if(!is.na(APD[[v]][i])) {
        traits$elements[[trait_name]][[v]] <- APD[[v]][i]
      }
    }
    if(APD$type[i] == "categorical") {
      categorical_tmp <- value_levels %>% filter(trait_cat == APD$trait[i]) %>% select(label, description)

      for(j in seq_along(1:nrow(categorical_tmp))) {
        traits$elements$trait[i]$allowed_values_levels[j] <- util_df_to_list(categorical_tmp[j,])
      }
    }
  }
}

yaml::write_yaml(list(traits=traits), "export/traits2.yml")

traits_old <- get_schema("config/traits.yml", I("traits"))
`

**Comment by dfalster** [comment]
hi @ehwenk 

This is easier is we reduce the amount of for loops and instead use natural list workflows

```
library(tidyverse)
library(austraits.build)

APD <- 
  read_csv("export/data/APD_traits.csv", show_col_types = FALSE) %>%
  mutate(
    description = ifelse(is.na(description), description_encoded, paste0(description_encoded, ";", description)),
    units = ifelse(str_starts(units, fixed("{")), sprintf("'%s'", units),units),
    Entity_URI = paste0("https://w3id.org/APD/traits/", identifier),
    allowed_values_levels = ifelse(type_x == "categorical", "add_in", NA),
  ) %>%
  select(trait,
         label,
         description,
         comments,
         type = type_x,
         units,
         allowed_values_min = min,
         allowed_values_max = max,
         allowed_values_levels,
         entity_URI = Entity_URI
         )

value_levels <- read_csv("export/data/APD_categorical_values.csv", show_col_types = FALSE) %>%
  select(trait_cat = trait_name,
         label,
         description
  )

traits <- get_schema("export/traits2.yml", I("traits"))

vars <- c("label", "description", "comments", "type", "units", "allowed_values_min", "allowed_values_max", "allowed_values_levels", "entity_URI")

traits$elements <-  
  APD %>%
  split(APD$trait) %>% # turn into list, 1 element for each row
  # reprocess rows into a list
  map(~ .x %>% 
        # variables in specified order
        select(all_of(vars)) %>%
        # remove NAs
        select(where(~ !is.na(.x))) %>%
        # as a list
        as.list()
  )

# for add categorical traits, add allowable values
for(trait in names(traits$elements)) {
  if(traits$elements[[trait]][["type"]] == "categorical") {      
    traits$elements[[trait]][["allowed_values_levels"]] <- 
      value_levels %>% 
      filter(trait_cat == trait) %>% 
      select(label, description) %>%
      arrange(label) %>% 
      pivot_wider(names_from = label, values_from = description) %>% 
      as.list()
  }
}

yaml::write_yaml(list(traits = traits), "config/traits.yml")
```

**Comment by ehwenk** [comment]
@dfalster  Yes this works! 1 last tweak, is it possible to reorder based on `Entity_URI`. Right now it is alphabetical by trait name, but it would be better to be sorted by URI's, because then similar traits near each other.

**Comment by dfalster** [comment]
Yep, so in that case just add a line to end of section loading APD to sort by UURI, i.e.

```
APD <- 
  read_csv("export/data/APD_traits.csv", show_col_types = FALSE) %>%
  .... %>%
  arrange(entity_URI)
```

**Comment by dfalster** [comment]
Ok, @ehwenk pointed out that still in wrong order. Here's a modification, with

- split by entity_URI (this causes output to be sorted by entity_URI)
- add trait into list as a double check everything's working. Remove once you've reviewed and happy 

```
library(tidyverse)
library(austraits.build)

APD <- 
  read_csv("export/data/APD_traits.csv", show_col_types = FALSE) %>%
  mutate(
    description = ifelse(is.na(description), description_encoded, paste0(description_encoded, ";", description)),
    units = ifelse(str_starts(units, fixed("{")), sprintf("'%s'", units),units),
    Entity_URI = paste0("https://w3id.org/APD/traits/", identifier),
    allowed_values_levels = ifelse(type_x == "categorical", "add_in", NA),
  ) %>%
  select(trait,
         label,
         description,
         comments,
         type = type_x,
         units,
         allowed_values_min = min,
         allowed_values_max = max,
         allowed_values_levels,
         entity_URI = Entity_URI
         ) %>%
  arrange(entity_URI)

value_levels <- read_csv("export/data/APD_categorical_values.csv", show_col_types = FALSE) %>%
  select(trait_cat = trait_name,
         label,
         description
  )

traits <- get_schema("export/traits2.yml", I("traits"))

vars <- c("trait", "label", "description", "comments", "type", "units", "allowed_values_min", "allowed_values_max", "allowed_values_levels", "entity_URI")

traits$elements <-  
  APD %>%
  split(APD$entity_URI) %>% # turn into list, 1 element for each row
  # reprocess rows into a list
  map(~ .x %>% 
        # variables in specified order
        select(all_of(vars)) %>%
        # remove NAs
        select(where(~ !is.na(.x))) %>%
        # as a list
        as.list()
  )

# rename with trait name
all.equal(names(traits$elements), APD$entity_URI)
names(traits$elements) <- APD$trait


# for add categorical traits, add allowable values
for(trait in names(traits$elements)) {
  if(traits$elements[[trait]][["type"]] == "categorical") {      
    traits$elements[[trait]][["allowed_values_levels"]] <- 
      value_levels %>% 
      filter(trait_cat == trait) %>% 
      select(label, description) %>%
      arrange(label) %>% 
      pivot_wider(names_from = label, values_from = description) %>% 
      as.list()
  }
}

yaml::write_yaml(list(traits = traits), "config/traits.yml")
```

**Comment by dfalster** [comment]
closed via #727 d2d408eda11a699404ae1570bf4ad2ecd11ba52a
